### PR TITLE
Ign/navaneeth blackhole moe hang

### PR DESCRIPTION
**Pull request by ign-navaneethk**
### Summary
Fix for hanging of ttnn.moe() kernel for input shape of  (2, 2, 32, 64) 
Fixes #39795 

#####  `ttnn/cpp/ttnn/operations/reduction/moe/device/kernels/dataflow/reader_create_index_tensor.cpp`:

- Pushes to `expert_mask_cb` but compute kernel currently does not implement expert mask. (Might be looked into in another issue)
- `topk_mask` follows the `add_tiles_bcast_rows` in compute kernel and does not belong to the `Ht` loop

##### `ttnn/cpp/ttnn/operations/reduction/moe/device/moe_program_factory.cpp`:

- Only defined CB depth of 4 tiles for `topk_values_cb` and `topk_indices_cb`. 
- Compute populates the `topk_values_cb` and `topk_indices_cb` with `Ht * Kt` tiles (Works for (1, 1, 32, 64) since `Ht*Kt < 4`)). 

##### `ttnn/cpp/ttnn/operations/reduction/moe/device/kernels/compute/moe.cpp`

- `eqz_tile()` was called without `copy_tile_to_dst()` which was giving wrong outputs when ran in loop for multiple tiles.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:ign/navaneeth_blackhole_moe_hang)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:ign/navaneeth_blackhole_moe_hang)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:ign/navaneeth_blackhole_moe_hang)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:ign/navaneeth_blackhole_moe_hang)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:ign/navaneeth_blackhole_moe_hang)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:ign/navaneeth_blackhole_moe_hang)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:ign/navaneeth_blackhole_moe_hang)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:ign/navaneeth_blackhole_moe_hang)

### CODE REVIEW

**Comment by ign-navaneethk** [comment]
@iwroszTT While optimising for cb depth, i have updated the reader kernel as well 
```
ttnn/cpp/ttnn/operations/reduction/moe/device/kernels/dataflow/reader_create_index_tensor.cpp
```
I think we can also remove the extra depth specified for the `uint32_t topk_mask_cb_units = Kt;` tensor as well if we bring the `cb_wait_front(in1_cb, cols);` inside the loop  in `add_block_bcast_rows_inplace()` in 
```
ttnn/cpp/ttnn/operations/reduction/moe/device/kernels/compute/moe.cpp
```

However this is not an issue currently since `k=32 -> Kt=1` for current kernel

File: tests/ttnn/nightly/unit_tests/operations/reduction/test_reduction_ops.py [hunk 1]
```diff
@@ -1004,13 +1004,13 @@ def test_moe(device, tensor_shape, dtype, layout):
         # Height is 1 so the mask broadcasts across all H rows.
         # Columns [E:] are -inf; adding this to the input ensures softmax
         # drives inactive expert probabilities to zero.
-        expert_mask = torch.zeros([N, C, 1, W], dtype=dtype)
+        expert_mask = torch.zeros([1, 1, 1, W], dtype=dtype)
```
Comment (bbradelTT) [inline tests/ttnn/nightly/unit_tests/operations/reduction/test_reduction_ops.py:1007]: Why is this changed?

File: ttnn/cpp/ttnn/operations/reduction/moe/device/kernels/dataflow/reader_create_index_tensor.cpp [hunk 2]
```diff
@@ -111,5 +91,29 @@ void kernel_main() {
...
+            generate_index_tile(cb_intermed_index, j);
+            noc.async_read(s0, cb_in0, tile_bytes_input, {.page_id = tile_id}, {.offset_bytes = tile_bytes_input});
+            tile_id++;
+            generate_index_tile(cb_intermed_index, j + 1);
+            noc.async_read_barrier();
+            cb_in0.push_back(2);
+        }
+    }
+
+    // topk mask uses add_tiles_bcast_rows() in compute so no need for Ht loop
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/reduction/moe/device/kernels/dataflow/reader_create_index_tensor.cpp]: Please rephrase to make this more clear.

File: ttnn/cpp/ttnn/operations/reduction/moe/device/kernels/dataflow/reader_create_index_tensor.cpp [hunk 3]
```diff
@@ -79,30 +79,10 @@ void kernel_main() {
...
-        for (uint32_t j = 0; j < Kt; ++j) {
-            noc.async_read(
-                s1, cb_topk, tile_bytes_topk, {.page_id = tile_id_topk}, {.offset_bytes = j * tile_bytes_topk});
-            tile_id_topk++;
-        }
-        noc.async_read_barrier();
-        cb_topk.push_back(Kt);
-
         // expert mask
+        /* Not consumed by compute kernel enable once support is added to compute kernel
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/reduction/moe/device/kernels/dataflow/reader_create_index_tensor.cpp]: What has been disabled?

File: ttnn/cpp/ttnn/operations/reduction/moe/device/moe_device_operation.cpp [hunk 4]
```diff
@@ -43,13 +46,31 @@ void MoeDeviceOperation::validate_on_program_cache_miss(
...
+        topk_logical_shape.rank(),
+        topk_logical_shape,
+        args.k);
     TT_FATAL(
-        expert_shape[-1] == input_shape[-1],
-        "Expert shape inner dim must be equal to input_shape[-1], got {}",
-        expert_shape[-1]);
-    TT_FATAL(topk_shape[-2] == 32, "Topk shape inner dim must be equal to 32, got {}", topk_shape[-2]);
-    TT_FATAL(expert_shape[-2] == 32, "Expert shape inner dim must be equal to 32, got {}", expert_shape[-2]);
+        is_row_broadcastable_mask(expert_logical_shape, input_logical_shape[-1]),
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/reduction/moe/device/moe_device_operation.cpp:68]: Would the doc string in nanobind need to be updated?

**Review by bbradelTT — COMMENTED** [review]
Good catch on the missing copy_tile.

How big do the inputs need to be before the CBs get too big for L1?

**Comment by ign-navaneethk** [comment]
> Good catch on the missing copy_tile.
> 
> How big do the inputs need to be before the CBs get too big for L1?

The number of height tiles should be the deciding factor since the kernel is limited to `Kt=1` for now. From quick test i can see anything above 120 tiles in height for bfloat16 will exceed L1 for Wormhole. So something upto (12, 10, 32, x) should be fine.

File: tests/ttnn/nightly/unit_tests/operations/reduction/test_reduction_ops.py (hunk 1 above)
Comment (ign-navaneethk) [reply tests/ttnn/nightly/unit_tests/operations/reduction/test_reduction_ops.py:1007]: The expert_mask and topk_mask tensors are row-broadcasted within the `add_block_bcast_rows_inplace()`. So the compute kernel was never using the values for `N, C > 1`.
New asserts have been added to `validate_on_program_cache_miss()` check for these shapes now.

File: ttnn/cpp/ttnn/operations/reduction/moe/device/kernels/dataflow/reader_create_index_tensor.cpp (hunk 3 above)
Comment (ign-navaneethk) [reply ttnn/cpp/ttnn/operations/reduction/moe/device/kernels/dataflow/reader_create_index_tensor.cpp]: [reader_create_index_tensor.cpp](https://github.com/tenstorrent/tt-metal/pull/43264/files/c30e8c53138a46b3926766cec0894f47b6563776#diff-a4ae7d55313f3836a1b75207dafd4f2578fdd2e34fd4575604ae16bc4806cb74)  used to push to the `expert_mask_cb`, but the compute had the expert_mask application step commented out 
```
    // mask out invalid experts
    // TODO: fix the bug that makes this give bad results
    // add_block_bcast_rows_inplace(input_cb_index, expert_mask_cb_index, Ht,Wt, true);
```
which was causing the cb to fill up and deadlock since all the input cb's were handled in the same loop. It was working for (1, 1, 32, 64) shape since the cb depth was set at 4 and the operation ended before it hit cb depth limit.

File: ttnn/cpp/ttnn/operations/reduction/moe/device/moe_device_operation.cpp (hunk 4 above)
Comment (ign-navaneethk) [reply ttnn/cpp/ttnn/operations/reduction/moe/device/moe_device_operation.cpp:68]: I have updated the doc string now:

> 
   - For the :attr:`topk_mask_tensor`, the logical shape must be row-broadcastable (N, C, H must be 1) 
       and the W must match :attr:`k` (i.e. 32). The padded shape H must be 32.
   - For the :attr:`expert_mask_tensor`, the logical shape must be row-broadcastable (N, C, H must be 1) 
       and the W must match W of the :attr:`input_tensor`. The padded shape H must be 32.


Have added back the removed checks as well since the new ones are checking the logical shape to verify broadcast behaviour. The original assert would assure shape is padded to full tile.

File: tests/ttnn/nightly/unit_tests/operations/reduction/test_reduction_ops.py [hunk 5]
```diff
@@ -1068,7 +1068,7 @@ def test_moe(device, tensor_shape, dtype, layout):
 
         return
 
-    atol = rtol = 0.01
+    atol = rtol = 0.02
```
Comment (fplavecTT) [inline tests/ttnn/nightly/unit_tests/operations/reduction/test_reduction_ops.py]: If looser tolerance is needed for only one shape, we should adjust it only for that one shape and leave stricter tolerance for other shapes.

File: ttnn/cpp/ttnn/operations/reduction/moe/device/kernels/dataflow/reader_create_index_tensor.cpp (hunk 3 above)
Comment (bbradelTT) [reply ttnn/cpp/ttnn/operations/reduction/moe/device/kernels/dataflow/reader_create_index_tensor.cpp]: Please update the comment to make it more clear.

File: ttnn/cpp/ttnn/operations/reduction/moe/device/kernels/dataflow/reader_create_index_tensor.cpp [hunk 6]
```diff
@@ -79,30 +79,10 @@ void kernel_main() {
...
-        for (uint32_t j = 0; j < Kt; ++j) {
-            noc.async_read(
-                s1, cb_topk, tile_bytes_topk, {.page_id = tile_id_topk}, {.offset_bytes = j * tile_bytes_topk});
-            tile_id_topk++;
-        }
-        noc.async_read_barrier();
-        cb_topk.push_back(Kt);
-
-        // expert mask
+        // Expert mask is not consumed by the compute kernel. Enable back once support is added to the compute kernel.
```
Comment (iwroszTT) [inline ttnn/cpp/ttnn/operations/reduction/moe/device/kernels/dataflow/reader_create_index_tensor.cpp:84]: Please update the doc string that expert mask is ignored.

Comment (iwroszTT) [reply ttnn/cpp/ttnn/operations/reduction/moe/device/kernels/dataflow/reader_create_index_tensor.cpp:84]: Please also update the docs example accordingly, and pre-apply the expert mask before calling ttnn::moe (as it is done in the pytest)
File: tests/ttnn/docs_examples/test_reduction_examples.py:189

**Review by bbradelTT — APPROVED** [review]
